lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -511,20 +598,92 @@ LogicalResult BuildOpGroups::buildOp(PatternRewriter &rewriter,
 
 LogicalResult BuildOpGroups::buildOp(PatternRewriter &rewriter,
                                      scf::YieldOp yieldOp) const {
-  if (yieldOp.getOperands().size() == 0)
+  if (yieldOp.getOperands().empty()) {
+    // If yiedOp operands are empty, we assume we have a for loop.
+    auto forOp = dyn_cast<scf::ForOp>(yieldOp->getParentOp());
+    assert(forOp);
```

Comment:
Why not return failure and emit errors instead of hard crashes? It is easier to debug than a hard crash.

Reply:
In addition to what is easier to debug, I think it's important to call out that there is an important distinction between assertions and errors. Assertions are to let compiler writers know something went wrong; they uphold invariants we expect within the implementation. Errors are to let users know something went wrong. Another way to look at this is if IR is legal (passes verifiers, etc.) assertions shouldn't fire in the passes--it is a bug of the pass for the compiler writer to fix if assertions fire.

Reply:
I guess my thinking was, since `while` loops have to yield something, the *only* time we would have empty yields is For Loops, so the parent really should be a ForOp. If we want to support an empty yield op when the parent is not a For Loop, we should probably look at/change this code, and the assert is a good reminder for that.